### ISSUE TITLE
`docker run` can't rely on the Events API.

### ISSUE THREAD

**Issue by jlhawn** [moby/moby#32242]
**Description**

Since Docker 1.13, the implementation of `docker run ...` changed to use the `Events` API over `ContainerWait` to wait for the container to exit. This was done to facilitate the move of handling `autoRemove` on the daemon rather than the CLI - the problem with the Wait API being that it only waited for the container to exit, and not for the container to be removed which would cause race conditions if, for example, a successive CLI command tried to create a container with the same name and it had not yet been cleaned up by the daemon.

While there is no problem with this Events API approach when connecting directly to the docker daemon 
which is running the container, I have seen several issue with it when the "daemon" is actually a Docker Swarm cluster (classic swarm, not swarmkit). One major problem is that swarm does not support event filtering at all, and so it ignores the container filter and ends up streaming back all cluster events. This means that your CLI may receive a "die" event for another container, causing your CLI to exit pre-maturely. Even if filtering were added to the swarm manager, there is yet another issue where its aggregated event stream is not reliable - it may temporarily lose the event stream from the engine which is running the target container but will not disconnect the client because the aggregate stream as a whole is still working. If the container exits during this short disconnection window, the events will never be sent to the CLI and it will hang forever. There is no way to guarantee that the event will be sent back to the client.

**Comment by jlhawn** [comment]
As an alternative, @tonistiigi once suggested a change to the `ContainerWait` API to support waiting until the container is removed. This would completely avoid the Event stream reliability issues with Docker Swarm. I've already started working on such changes here: https://github.com/docker/docker/pull/32237

**Comment by jlhawn** [comment]
@thaJeztah Can you reopen this issue, please? It's not really supposed to be closed yet since the `cli` package was removed. I'll need to make another PR here: https://github.com/docker/cli

**Comment by jlhawn** [comment]
Or maybe I should just move this issue there

**Comment by thaJeztah** [comment]
Let me reopen for visibility, help me remind to close it once the cli PR is merged

**Comment by thaJeztah** [comment]
this was moved to / continued in https://github.com/docker/cli/issues/100 and fixed by https://github.com/docker/cli/pull/101
